nexus/src/db/queries/ip_pool.rs

```diff
@@ -0,0 +1,288 @@
+) -> QueryResult<()> {
+    out.push_sql("SELECT ");
+    out.push_identifier(dsl::id::NAME)?;
+    out.push_sql(" FROM ");
+    IP_POOL_RANGE_FROM_CLAUSE.walk_ast(out.reborrow())?;
+    out.push_sql(" WHERE ");
+    out.push_bind_param::<sql_types::Inet, IpNetwork>(address)?;
+    out.push_sql(" >= ");
+    out.push_identifier(dsl::first_address::NAME)?;
+    out.push_sql(" AND ");
+    out.push_bind_param::<sql_types::Inet, IpNetwork>(address)?;
+    out.push_sql(" <= ");
+    out.push_identifier(dsl::last_address::NAME)?;
+    out.push_sql(" AND ");
+    out.push_identifier(dsl::time_deleted::NAME)?;
+    out.push_sql(" IS NULL LIMIT 1");
+    Ok(())
+}
+
+// Push the subquery that finds any records with an address contained within the
+// provided candidate range.
+//
+// ```sql
+// SELECT
+//      id
+// FROM
+//      ip_pool_range
+// WHERE
+//      <column> >= <first_address> AND
+//      <column> <= <last_address> AND
+//      time_deleted IS NULL
+// LIMIT 1
+// ```
+fn push_candidate_contains_record_subquery<'a, C>(
+    mut out: AstPass<'_, 'a, Pg>,
+    first_address: &'a IpNetwork,
+    last_address: &'a IpNetwork,
+) -> QueryResult<()>
+where
+    C: Column<Table = dsl::ip_pool_range>,
+{
+    out.push_sql("SELECT ");
+    out.push_identifier(dsl::id::NAME)?;
+    out.push_sql(" FROM ");
+    IP_POOL_RANGE_FROM_CLAUSE.walk_ast(out.reborrow())?;
+    out.push_sql(" WHERE ");
+    out.push_identifier(C::NAME)?;
+    out.push_sql(" >= ");
+    out.push_bind_param::<sql_types::Inet, IpNetwork>(first_address)?;
+    out.push_sql(" AND ");
+    out.push_identifier(C::NAME)?;
+    out.push_sql(" <= ");
+    out.push_bind_param::<sql_types::Inet, IpNetwork>(last_address)?;
+    out.push_sql(" AND ");
+    out.push_identifier(dsl::time_deleted::NAME)?;
+    out.push_sql(" IS NULL LIMIT 1");
+    Ok(())
+}
+
+impl QueryFragment<Pg> for FilterOverlappingIpRanges {
+    fn walk_ast<'a>(&'a self, mut out: AstPass<'_, 'a, Pg>) -> QueryResult<()> {
+        out.unsafe_to_cache_prepared();
+        out.push_sql("SELECT ");
+        out.push_bind_param::<sql_types::Uuid, Uuid>(&self.range.id)?;
+        out.push_sql(", ");
+        out.push_bind_param::<sql_types::Timestamptz, DateTime<Utc>>(
+            &self.range.time_created,
+        )?;
+        out.push_sql(", ");
+        out.push_bind_param::<sql_types::Timestamptz, DateTime<Utc>>(
+            &self.range.time_modified,
+        )?;
+        out.push_sql(", ");
+        out.push_bind_param::< sql_types::Nullable<sql_types::Timestamptz>, Option<DateTime<Utc>>>(&self.range.time_deleted)?;
+        out.push_sql(", ");
+        out.push_bind_param::<sql_types::Inet, IpNetwork>(
+            &self.range.first_address,
+        )?;
+        out.push_sql(", ");
+        out.push_bind_param::<sql_types::Inet, IpNetwork>(
+            &self.range.last_address,
+        )?;
+        out.push_sql(", ");
+        out.push_bind_param::<sql_types::Uuid, Uuid>(&self.range.ip_pool_id)?;
```

Comment:
I'm with you. Every time I've tried to just push a struct as a SQL type, I get into a morass of compiler errors in Diesel. I agree that it _should_ work, and yes it'll make this query (and many others!) robust to future changes. I'll try it again.

Reply:
Ok, I think this unfortunately does not work. `ToSql` is for serializing _single_ values to the database, not a sequence of values like an entire record. The generated `SqlType` is indeed a tuple of all the columns, but that's not really something we can convert to a single expression. I'm going to defer this for now, but given how many times we've done this already, it may be worth us adding our own trait (and probably a derive-macro) that supports selecting an instance of a struct like this as a constant.

For future reference, I think we'll need to implement [`IntoSql`](https://docs.diesel.rs/master/diesel/expression/trait.IntoSql.html) for doing things like `table.select(my_model_type.into_sql())`; and some other method that looks like: `fn push_as_bind_group(&self, mut out: diesel::query_builder::AstPass) -> diesel::QueryResult<()>`, which calls `out.push_bind_param` for each field in the struct.